src/Generator/TypeGenerator.php

```diff
@@ -73,6 +78,32 @@ public static function fromTypeString($type)
         return $instance;
     }
 
+    /**
+     *  @param array $typeArray
+     *
+     *  @return TypeGenerator
+     *
+     *  @throws InvalidArgumentException
+     */
+    public static function fromTypeArray(array $typeArray)
```

Comment:
I'd avoid this overall. This forces compatibility on array keys: just make it multiple parameters.